File: app/src/main/java/org/sorz/lab/tinykeepass/EntryFragment.java
package org.sorz.lab.tinykeepass;

import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Bundle;
import android.app.Fragment;
import android.os.SystemClock;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import de.slackspace.openkeepass.domain.Entry;


public class EntryFragment extends Fragment implements SearchView.OnQueryTextListener {
    private static final long INACTIVE_AUTO_LOCK_MILLIS = 3 * 60 * 1000;

    private MainActivity activity;
    private EntryRecyclerViewAdapter entryAdapter;
    private ClipboardManager clipboardManager;
    private LocalBroadcastManager localBroadcastManager;
    private FloatingActionButton fab;
    private ActionMode actionMode;
    private long lastPauseTimeMillis;

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public EntryFragment() {
    }


    public static EntryFragment newInstance() {
        return new EntryFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_entry_list, container, false);

        // Set the adapter

        Context context = view.getContext();
        RecyclerView recyclerView = view.findViewById(R.id.list);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        entryAdapter = new EntryRecyclerViewAdapter(
                getContext(), this::onEntryClick, this::onEntryLongClick);
        recyclerView.setAdapter(entryAdapter);

        fab = view.findViewById(R.id.fab);
        fab.setOnClickListener(v -> {
            if (activity != null) {
                fab.hide();
                activity.doSyncDatabase();
            }
        });
        return view;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        activity = (MainActivity) context;
        localBroadcastManager = LocalBroadcastManager.getInstance(activity);
        clipboardManager = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        activity = null;
    }

    @Override
    public void onStart() {
        super.onStart();
        localBroadcastManager.registerReceiver(broadcastReceiver,
                new IntentFilter(DatabaseSyncingService.BROADCAST_SYNC_FINISHED));
    }

    @Override
    public void onStop() {
        super.onStop();
        localBroadcastManager.unregisterReceiver(broadcastReceiver);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (KeePassStorage.get() == null ||
                lastPauseTimeMillis > 0 &&
                        SystemClock.elapsedRealtime() - lastPauseTimeMillis >
                                INACTIVE_AUTO_LOCK_MILLIS) {
            activity.doLockDatabase();
            activity.doUnlockDatabase();
       }
    }

    @Override
    public void onPause() {
        super.onPause();
        lastPauseTimeMillis = SystemClock.elapsedRealtime();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_main, menu);
        MenuItem searchMenu = menu.findItem(R.id.action_search);
        SearchView searchView = (SearchView) searchMenu.getActionView();
        searchView.setOnQueryTextListener(this);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_lock_db:
                if (activity != null)
                    activity.doLockDatabase();
                return true;
            case R.id.action_exit:
                if (activity != null)
                    getActivity().finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return onQueryTextChange(query);
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        entryAdapter.setFilter(newText);
        return true;
    }

    /**
     * Immediately copy password & show count down notification.
     * @param entry to copy
     */
    private void copyPassword(Entry entry) {
        Intent intent = new Intent(getContext(), PasswordCopingService.class);
        intent.setAction(PasswordCopingService.ACTION_COPY_PASSWORD);
        intent.putExtra(PasswordCopingService.EXTRA_PASSWORD, entry.getPassword());
        getContext().startService(intent);
        if (getView() != null) {
            Snackbar snackbar = Snackbar.make(getView(), R.string.password_copied, Snackbar.LENGTH_SHORT);
            snackbar.setAction(R.string.show_password, view -> {
                showPassword(entry);

                // Cancel password copying
                Intent cancelIntent = new Intent(getContext(), PasswordCopingService.class);
                cancelIntent.setAction(PasswordCopingService.ACTION_CLEAN_CLIPBOARD);
                getContext().startService(cancelIntent);
            });
            snackbar.show();
        } else {
            Toast.makeText(getContext(), R.string.password_copied, Toast.LENGTH_SHORT).show();
        }
    }

    /**
     * Display password.
     * @param entry to show
     */
    private void showPassword(Entry entry) {
        if (getActivity() == null)
            return;
        if (actionMode != null) {
            actionMode.finish();
        }
        actionMode = getActivity().startActionMode(entryShowPasswordActionModeCallback);
        if (actionMode != null) {
            actionMode.setTag(entryShowPasswordActionModeCallback);
            actionMode.setTitle(getString(R.string.title_show_password));
            entryAdapter.showPassword(entry);
        }
    }

    private void copyEntry(Entry entry, boolean copyUsername, boolean copyPassword) {
        if (copyUsername && notEmpty(entry.getUsername())) {
            clipboardManager.setPrimaryClip(
                    ClipData.newPlainText(getString(R.string.username), entry.getUsername()));
            String message = getString(R.string.username_copied, entry.getUsername());
            if (getView() == null) {
                Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
            } else {
                Snackbar snackbar = Snackbar.make(getView(), message, Snackbar.LENGTH_LONG);
                if (copyPassword)
                    snackbar.setAction(R.string.copy_password,v -> copyPassword(entry));
                snackbar.show();
            }
        }
        if (copyPassword && notEmpty(entry.getPassword())) {
            if (copyUsername && notEmpty(entry.getUsername())) {
                // username already copied, waiting for user's action before copy password.
                Intent intent = new Intent(getContext(), PasswordCopingService.class);
                intent.setAction(PasswordCopingService.ACTION_NEW_NOTIFICATION);
                intent.putExtra(PasswordCopingService.EXTRA_PASSWORD, entry.getPassword());
                if (entry.getUsername() != null)
                    intent.putExtra(PasswordCopingService.EXTRA_USERNAME, entry.getUsername());
                if (entry.getTitle() != null)
                    intent.putExtra(PasswordCopingService.EXTRA_ENTRY_TITLE, entry.getTitle());
                getContext().startService(intent);
            } else {
                // username not copied, copy password immediately.
                copyPassword(entry);
            }
        }
    }

    private void openEntryUrl(Entry entry) {
        if (entry.getUrl() == null || entry.getUrl().isEmpty())
            return;
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(entry.getUrl()));
        startActivity(intent);
        copyEntry(entry, true, true);
    }

    private void onEntryClick(View view, Entry entry) {
        if (actionMode != null) {
            actionMode.finish();
        } else {
            copyEntry(entry, true, true);
        }
    }

    private boolean onEntryLongClick(View view, Entry entry) {
        if (getActivity() == null)
            return false;
        if (actionMode != null) {
            if (actionMode.getTag() == entryLongClickActionModeCallback) {
                actionMode.invalidate();
                return true;
            } else {
                // Finish show password mode
                actionMode.finish();
            }
        }
        actionMode = getActivity().startActionMode(entryLongClickActionModeCallback);
        if (actionMode != null) {
            actionMode.setTag(entryLongClickActionModeCallback);
            return true;
        }
        return false;
    }

    private boolean notEmpty(String string) {
        return string != null && !string.isEmpty();
    }

    private ActionMode.Callback entryShowPasswordActionModeCallback = new ActionMode.Callback() {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem menuItem) {
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            actionMode = null;
            entryAdapter.hidePassword();
        }
    };


    private ActionMode.Callback entryLongClickActionModeCallback = new ActionMode.Callback() {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.entry_context, menu);
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            Entry entry = entryAdapter.getSelectedItem();
            if (entry == null)
                return false;
            menu.findItem(R.id.action_copy_username).setVisible(notEmpty(entry.getUsername()));
            menu.findItem(R.id.action_copy_password).setVisible(notEmpty(entry.getPassword()));
            menu.findItem(R.id.action_copy_url).setVisible(notEmpty(entry.getUrl()));
            menu.findItem(R.id.action_open).setVisible(notEmpty(entry.getUrl()));
            return true;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            Entry entry = entryAdapter.getSelectedItem();
            if (entry == null)
                return false;
            switch (item.getItemId()) {
                case R.id.action_copy_username:
                    copyEntry(entry, true, false);
                    break;
                case R.id.action_copy_password:
                    copyEntry(entry, false, true);
                    break;
                case R.id.action_show_password:
                    mode.finish();
                    showPassword(entry);
                    break;
                case R.id.action_copy_url:
                    if (notEmpty(entry.getUrl())) {
                        clipboardManager.setPrimaryClip(
                                ClipData.newPlainText("URL", entry.getUrl()));
                        if (getView() != null)
                            Snackbar.make(getView(), R.string.url_copied,
                                    Snackbar.LENGTH_SHORT).show();
                    }
                    break;
                case R.id.action_open:
                    openEntryUrl(entry);
                    break;
                default:
                    return false;
            }
            mode.finish();
            return true;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            actionMode = null;
            entryAdapter.clearSelection();
        }
    };

        private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (DatabaseSyncingService.BROADCAST_SYNC_FINISHED.equals(intent.getAction())) {
                fab.show();
                String error = intent.getStringExtra(DatabaseSyncingService.EXTRA_SYNC_ERROR);

                if (error != null) {
                    if (getView() != null)
                        Snackbar.make(getView(), getString(R.string.fail_to_sync, error),
                                Snackbar.LENGTH_LONG).show();
                } else {
                    if (getView() != null)
                        Snackbar.make(getView(), R.string.sync_done, Snackbar.LENGTH_SHORT).show();
                    entryAdapter.reloadEntries();
                }
            }
        }
    };
}

File: app/src/main/java/org/sorz/lab/tinykeepass/EntryFragment.kt
package org.sorz.lab.tinykeepass

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.SystemClock
import com.google.android.material.snackbar.Snackbar
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager.LayoutParams.FLAG_SECURE
import android.widget.Toast
import androidx.work.WorkInfo.State
import androidx.work.WorkManager
import com.google.android.material.floatingactionbutton.FloatingActionButton

import org.sorz.lab.tinykeepass.keepass.KeePassStorage

import com.kunzisoft.keepass.database.element.Entry


private const val INACTIVE_AUTO_LOCK_MILLIS = (3 * 60 * 1000).toLong()


class EntryFragment : BaseEntryFragment() {
    override val fragmentLayout: Int = R.layout.fragment_entry_list
    private lateinit var fab: FloatingActionButton

    private val clipboardManager by lazy(LazyThreadSafetyMode.NONE) {
        requireContext().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    }
    private var actionMode: ActionMode? = null
    private var lastPauseTimeMillis: Long = 0

    private val entryShowPasswordActionModeCallback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean = true
        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean = false
        override fun onActionItemClicked(mode: ActionMode, menuItem: MenuItem): Boolean = false
        override fun onDestroyActionMode(mode: ActionMode) {
            actionMode = null
            hidePassword()
        }
    }

    private val entryLongClickActionModeCallback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.entry_context, menu)
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            val entry = entryAdapter.selectedEntry ?: return false
            menu.run {
                findItem(R.id.action_copy_username).isVisible = !entry.username.isBlank()
                findItem(R.id.action_copy_password).isVisible = !entry.password.isBlank()
                findItem(R.id.action_copy_url).isVisible = !entry.url.isBlank()
                findItem(R.id.action_open).isVisible = !entry.url.isBlank()
            }
            return true
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            val entry = entryAdapter.selectedEntry ?: return false
            when (item.itemId) {
                R.id.action_copy_username -> copyEntry(entry, true, false)
                R.id.action_copy_password -> copyEntry(entry, false, true)
                R.id.action_show_password -> {
                    mode.finish()
                    showPassword(entry)
                }
                R.id.action_copy_url -> if (!entry.url.isBlank()) {
                    clipboardManager.setPrimaryClip(ClipData.newPlainText("URL", entry.url))
                    if (view != null)
                        Snackbar.make(view!!, R.string.url_copied,
                                Snackbar.LENGTH_SHORT).show()
                }
                R.id.action_open -> openEntryUrl(entry)
                else -> return false
            }
            mode.finish()
            return true
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            actionMode = null
            entryAdapter.clearSelection()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)!!
        fab = view.findViewById(R.id.fab)
        fab.setOnClickListener {
            activity.run {
                if (this is MainActivity) doSyncDatabase()
            }
        }

        WorkManager.getInstance(requireContext())
            .getWorkInfosForUniqueWorkLiveData(DATABASE_SYNCING_WORK_NAME)
            .observe(viewLifecycleOwner) { works ->
                val work = works.lastOrNull()
                val msg = when (work?.state) {
                    null -> return@observe fab.show()
                    State.SUCCEEDED -> getString(R.string.sync_done)
                    State.CANCELLED -> getString(R.string.sync_cancelled)
                    State.FAILED -> getString(
                        R.string.fail_to_sync,
                        work.outputData.getString(RESULT_ERROR) ?: "I/O error"
                    )
                    State.RUNNING, State.BLOCKED, State.ENQUEUED -> return@observe fab.hide()
                }
                fab.show()
                val time = work.outputData.getLong(RESULT_TIMESTAMP, 0)
                if (System.currentTimeMillis() - time > 10 * 1000) return@observe
                val length = if (work.state == State.FAILED) Snackbar.LENGTH_LONG
                    else Snackbar.LENGTH_SHORT
                Snackbar.make(view, msg, length).show()
            }

        return view
    }

    override fun onResume() {
        super.onResume()
        if (KeePassStorage.get(context) == null || lastPauseTimeMillis > 0
                && SystemClock.elapsedRealtime() - lastPauseTimeMillis > INACTIVE_AUTO_LOCK_MILLIS) {
            // time outed, lock and exit to unlock dialog.
            (activity as MainActivity).run {
                doLockDatabase()
                Handler().post { doUnlockDatabase() }
            }
        }
    }

    override fun onPause() {
        super.onPause()
        lastPauseTimeMillis = SystemClock.elapsedRealtime()
        if (actionMode?.tag == entryShowPasswordActionModeCallback)
            actionMode?.finish()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_main, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_lock_db -> {
            (activity as MainActivity).doLockDatabase()
            true
        }
        R.id.action_exit -> {
            requireActivity().finish()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    /**
     * Immediately copy password & show count down notification.
     * @param entry to copy
     */
    private fun copyPassword(entry: Entry) {
        Intent(context, PasswordCopingService::class.java).apply {
            action = PasswordCopingService.ACTION_COPY_PASSWORD
            putExtra(PasswordCopingService.EXTRA_PASSWORD, entry.password)
            requireContext().startService(this)
        }
        view?.also { view ->
            Snackbar.make(view, R.string.password_copied, Snackbar.LENGTH_SHORT).show()
        } ?: Toast.makeText(context, R.string.password_copied, Toast.LENGTH_SHORT).show()
    }

    /**
     * Display password.
     * @param entry to show
     */
    private fun showPassword(entry: Entry) {
        requireActivity().window.setFlags(FLAG_SECURE, FLAG_SECURE)
        actionMode?.finish()
        actionMode = requireActivity().startActionMode(entryShowPasswordActionModeCallback)?.apply {
            tag = entryShowPasswordActionModeCallback
            title = getString(R.string.title_show_password)
            entryAdapter.showPassword(entry)
        }
    }

    private fun hidePassword() {
        entryAdapter.hidePassword()
        requireActivity().window.setFlags(0, FLAG_SECURE)
    }

    private fun copyEntry(entry: Entry, copyUsername: Boolean, copyPassword: Boolean) {
        if (copyUsername && entry.username.isNotEmpty()) {
            ClipData.newPlainText(getString(R.string.username), entry.username).apply {
                clipboardManager.setPrimaryClip(this)
            }
            val message = getString(R.string.username_copied, entry.username)
            view.also { view ->
                Snackbar.make(view!!, message, Snackbar.LENGTH_LONG).apply {
                    if (copyPassword) setAction(R.string.copy_password) { copyPassword(entry) }
                    show()
                }
            }
        }
        if (copyPassword && entry.password.isNotEmpty()) {
            if (copyUsername && entry.username.isNotEmpty()) {
                // username already copied, waiting for user's action before copy password.
                val intent = Intent(context, PasswordCopingService::class.java).apply {
                    action = PasswordCopingService.ACTION_NEW_NOTIFICATION
                    putExtra(PasswordCopingService.EXTRA_PASSWORD, entry.password)
                    entry.username.also { username ->
                        putExtra(PasswordCopingService.EXTRA_USERNAME, username)
                    }
                    entry.title.also { title ->
                        putExtra(PasswordCopingService.EXTRA_ENTRY_TITLE, title)
                    }
                }
                requireContext().startService(intent)
            } else {
                // username not copied, copy password immediately.
                copyPassword(entry)
            }
        }
    }

    private fun openEntryUrl(entry: Entry) {
        if (entry.url.isEmpty())
            return
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(entry.url))
        startActivity(intent)
        copyEntry(entry, true, true)
    }

    override fun onEntryClick(view: View, entry: Entry) {
        actionMode?.run { finish() }
                ?: copyEntry(entry, true, true)
    }

    override fun onEntryLongClick(view: View, entry: Entry): Boolean {
        actionMode?.run {
            if (tag == entryLongClickActionModeCallback) {
                invalidate()
                return true
            } else {
                finish()
            }
        }

        actionMode = requireActivity().startActionMode(entryLongClickActionModeCallback)?.apply {
            tag = entryLongClickActionModeCallback
        } ?: return false
        return true
    }

    companion object {
        fun newInstance(): EntryFragment = EntryFragment()
    }
}
